src/vm/diagnosticprotocolhelper.cpp

```diff
@@ -35,55 +73,36 @@ void DiagnosticProtocolHelper::GenerateCoreDump(IpcStream* pStream)
     if (pStream == nullptr)
         return;
 
-    HRESULT hr = S_OK;
-
-    // TODO: Read within a loop.
-    uint8_t buffer[IpcStreamReadBufferSize] { };
-    uint32_t nNumberOfBytesRead = 0;
-    bool fSuccess = pStream->Read(buffer, sizeof(buffer), nNumberOfBytesRead);
-    if (fSuccess)
+    const GenerateCoreDumpCommandPayload* payload = message.TryParsePayload<GenerateCoreDumpCommandPayload>();
+    if (payload == nullptr)
     {
-        // The protocol buffer is defined as:
-        //   string - dumpName (UTF16)
-        //   int - dumpType
-        //   int - diagnostics
-        // returns
-        //   ulong - status
-        LPCWSTR pwszDumpName;
-        INT dumpType;
-        INT diagnostics;
-
-        uint8_t *pBufferCursor = buffer;
-        uint32_t bufferLen = nNumberOfBytesRead;
+        DiagnosticsIpc::IpcMessage errorMessage(MiscellaneousErrorHeader, DiagnosticsIpc::ServerErrorPayload{ DiagnosticsIpc::DiagnosticServerErrorCode::BadEncoding });
+        errorMessage.Send(pStream);
+        delete pStream;
+        return;
+    }
 
-        if (TryParseString(pBufferCursor, bufferLen, pwszDumpName) &&
-            TryParse(pBufferCursor, bufferLen, dumpType) &&
-            TryParse(pBufferCursor, bufferLen, diagnostics))
-        {
-            MAKE_UTF8PTR_FROMWIDE_NOTHROW(szDumpName, pwszDumpName);
-            if (szDumpName != nullptr)
-            {
-                if (!PAL_GenerateCoreDump(szDumpName, dumpType, diagnostics))
-                {
-                    hr = E_FAIL;
-                }
-            }
-            else 
-            {
-                hr = E_OUTOFMEMORY;
-            }
-        }
-        else
+    MAKE_UTF8PTR_FROMWIDE_NOTHROW(szDumpName, payload->dumpName);
+    if (szDumpName != nullptr)
+    {
+        if (!PAL_GenerateCoreDump(szDumpName, payload->dumpType, payload->diagnostics))
         {
-            hr = E_INVALIDARG;
+            DiagnosticsIpc::IpcMessage errorMessage(MiscellaneousErrorHeader, DiagnosticsIpc::ServerErrorPayload{ DiagnosticsIpc::DiagnosticServerErrorCode::BAD });
+            errorMessage.Send(pStream);
+            delete pStream;
+            return;
         }
     }
     else 
     {
-        hr = E_UNEXPECTED;
+        DiagnosticsIpc::IpcMessage errorMessage(MiscellaneousErrorHeader, DiagnosticsIpc::ServerErrorPayload{ DiagnosticsIpc::DiagnosticServerErrorCode::BAD });
+        errorMessage.Send(pStream);
+        delete pStream;
```

Comment:
Need to delete payload as well.

Reply:
In fact, I don't think you ever freed the payload. Probably want to free it in rest of the cases.

Reply:
Updated all code paths to delete the payload in error and success cases (except for `EventPipeProtocolHelper::CollectTracing` since the payload is passed to `EventPipe::Enable`)